core/src/main/java/com/linecorp/armeria/server/Http2KeepAliveHandler.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Stopwatch;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.EventLoop;
+import io.netty.handler.codec.http2.Http2Error;
+import io.netty.handler.codec.http2.Http2Exception;
+import io.netty.handler.codec.http2.Http2FrameWriter;
+import io.netty.handler.codec.http2.Http2PingFrame;
+import io.netty.handler.timeout.IdleState;
+import io.netty.handler.timeout.IdleStateEvent;
+import io.netty.handler.timeout.IdleStateHandler;
+import io.netty.util.concurrent.GenericFutureListener;
+
+/**
+ * This will send {@link Http2PingFrame} when an {@link IdleStateEvent} is emitted by {@link IdleStateHandler}.
+ * Specifically, it will write a PING frame to remote and then expects an ACK back within
+ * configured pingTimeOut. If timeout exceeds then channel will be closed.
+ *</p>
+ * This constructor will fail to initialize when pipeline does not have {@link IdleStateHandler}.
+ *</p>
+ * This class is not thread-safe and all methods are to be called from single thread such as {@link EventLoop}.
+ */
+@NotThreadSafe
+class Http2KeepAliveHandler {
+
+    private static final Logger logger = LoggerFactory.getLogger(Http2KeepAliveHandler.class);
+
+    private final Http2FrameWriter frameWriter;
+    private final ThreadLocalRandom random = ThreadLocalRandom.current();
+    private final Stopwatch stopwatch = Stopwatch.createUnstarted();
+
+    @Nullable
+    private ChannelFuture pingWriteFuture;
+    @Nullable
+    private Future<?> shutdownFuture;
+    private long pingTimeoutInNanos;
+    private State state = State.IDLE;
+    private Channel channel;
+    private long lastPingPayload;
+
+    private final Runnable shutdownRunnable = () -> {
+        logger.debug("Closing channel: {} as PING timed out.", channel);
```

Comment:
Our logging convention is to put the current channel or context at the first place, e.g.

    logger.debug("{} Closing channel due to PING timeout", channel);